esp32s3-hal/examples/embassy_twai.rs

```diff
@@ -0,0 +1,159 @@
+//! This example demonstrates use of the twai peripheral running the embassy
+//! executor and asynchronously receiving and transmitting twai frames.
+//!
+//! The `receiver` task waits to receive a frame and puts it into a channel
+//! which will be picked up by the `transmitter` task.
+//!
+//! The `transmitter` task waits for a channel to receive a frame and transmits
+//! it.
+//!
+//! This example should work with another ESP board running the `twai` example
+//! with `IS_SENDER` set to `true`.
+
+#![no_std]
+#![no_main]
+#![feature(type_alias_impl_trait)]
+
+use embassy_executor::Spawner;
+use embassy_sync::{blocking_mutex::raw::NoopRawMutex, channel::Channel};
+use embedded_can::{Frame, Id};
+use esp32s3_hal::{
+    clock::ClockControl,
+    embassy,
+    interrupt,
+    peripherals::{self, Peripherals, TWAI0},
+    prelude::*,
+    twai::{self, EspTwaiFrame, TwaiRx, TwaiTx},
+    IO,
+};
+use esp_backtrace as _;
+use esp_println::println;
+use static_cell::make_static;
+
+type TwaiOutbox = Channel<NoopRawMutex, EspTwaiFrame, 16>;
+
+#[embassy_executor::task]
+async fn receiver(mut rx: TwaiRx<'static, TWAI0>, channel: &'static TwaiOutbox) -> ! {
+    loop {
+        let frame = rx.receive_async().await;
+
+        match frame {
+            Ok(frame) => {
+                println!("Received a frame:");
+                print_frame(&frame);
+
+                // repeat the frame back
+                channel.send(frame).await;
+            }
+            Err(e) => {
+                println!("Receive error: {:?}", e);
+            }
+        }
+    }
+}
+
+#[embassy_executor::task]
+async fn transmitter(mut tx: TwaiTx<'static, TWAI0>, channel: &'static TwaiOutbox) -> ! {
+    loop {
+        let frame = channel.receive().await;
+        let result = tx.transmit_async(&frame).await;
+
+        match result {
+            Ok(()) => {
+                println!("Transmitted a frame:");
+                print_frame(&frame);
+            }
+            Err(e) => {
+                println!("Transmit error: {:?}", e);
+            }
+        }
+    }
+}
+
+#[main]
+async fn main(spawner: Spawner) {
+    let peripherals = Peripherals::take();
+    let system = peripherals.SYSTEM.split();
+    let clocks = ClockControl::boot_defaults(system.clock_control).freeze();
+
+    #[cfg(feature = "embassy-time-systick")]
+    embassy::init(
+        &clocks,
+        esp32s3_hal::systimer::SystemTimer::new(peripherals.SYSTIMER),
+    );
+
+    #[cfg(feature = "embassy-time-timg0")]
+    {
+        let timer_group0 = esp32s3_hal::timer::TimerGroup::new(peripherals.TIMG0, &clocks);
+        embassy::init(&clocks, timer_group0.timer0);
+    }
+
+    let io = IO::new(peripherals.GPIO, peripherals.IO_MUX);
+
+    // Use GPIO pins 2 and 3 to connect to the respective pins on the CAN
+    // transceiver.
+    let can_tx_pin = io.pins.gpio2;
```

Comment:
Can we add the open drain work around here with a comment to remove when using transceivers, like we do in the TWAI example?

Reply:
I haven't gotten the open drain bit to work, but now with the S3 I'll try again. Makes sense to include. We'll need the seed transmission like you said to make it work.